File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/receivers/notification/PlaybackNotificationRouter.java
package com.lasthopesoftware.bluewater.client.playback.service.receivers.notification;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;
import com.lasthopesoftware.bluewater.client.playback.service.notification.NotifyOfPlaybackEvents;
import com.vedsoft.futures.runnables.OneParameterAction;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;


public class PlaybackNotificationRouter extends BroadcastReceiver {

	private final Map<String, OneParameterAction<Intent>> mappedEvents;

	private final NotifyOfPlaybackEvents playbackNotificationBroadcaster;

	public PlaybackNotificationRouter(NotifyOfPlaybackEvents playbackNotificationBroadcaster) {
		this.playbackNotificationBroadcaster = playbackNotificationBroadcaster;

		mappedEvents = new HashMap<>(4);
		mappedEvents.put(PlaylistEvents.onPlaylistChange, this::onPlaylistChange);
		mappedEvents.put(PlaylistEvents.onPlaylistPause, i -> playbackNotificationBroadcaster.notifyPaused());
		mappedEvents.put(PlaylistEvents.onPlaylistStart, i -> playbackNotificationBroadcaster.notifyPlaying());
		mappedEvents.put(PlaylistEvents.onPlaylistStop, i -> playbackNotificationBroadcaster.notifyStopped());
	}

	public Set<String> registerForIntents() {
		return mappedEvents.keySet();
	}

	@Override
	public void onReceive(Context context, Intent intent) {
		final String action = intent.getAction();
		if (action == null) return;

		final OneParameterAction<Intent> eventHandler = mappedEvents.get(action);
		if (eventHandler != null)
			eventHandler.runWith(intent);
	}

	private void onPlaylistChange(Intent intent) {
		final int fileKey = intent.getIntExtra(PlaylistEvents.PlaybackFileParameters.fileKey, -1);
		if (fileKey >= 0)
			playbackNotificationBroadcaster.notifyPlayingFileChanged(new ServiceFile(fileKey));
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/receivers/notification/PlaybackNotificationRouter.kt
package com.lasthopesoftware.bluewater.client.playback.service.receivers.notification

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.properties.storage.FilePropertiesUpdatedMessage
import com.lasthopesoftware.bluewater.client.connection.libraries.ProvideScopedUrlKey
import com.lasthopesoftware.bluewater.client.playback.nowplaying.storage.GetNowPlayingState
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.PlaybackMessage
import com.lasthopesoftware.bluewater.client.playback.service.notification.NotifyOfPlaybackEvents
import com.lasthopesoftware.bluewater.shared.UrlKeyHolder
import com.lasthopesoftware.bluewater.shared.cls
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessage
import com.lasthopesoftware.bluewater.shared.messages.application.RegisterForApplicationMessages

class PlaybackNotificationRouter(
	private val playbackNotificationBroadcaster: NotifyOfPlaybackEvents,
	private val registerApplicationMessages: RegisterForApplicationMessages,
	private val scopedUrlKeys: ProvideScopedUrlKey,
	private val nowPlayingProvider: GetNowPlayingState,
) :
	(ApplicationMessage) -> Unit,
	AutoCloseable
{
	init {
		registerApplicationMessages.registerForClass(cls<PlaybackMessage.TrackChanged>(), this)
		registerApplicationMessages.registerForClass(cls<PlaybackMessage.PlaybackStarted>(), this)
		registerApplicationMessages.registerForClass(cls<PlaybackMessage.PlaybackPaused>(), this)
		registerApplicationMessages.registerForClass(cls<PlaybackMessage.PlaybackInterrupted>(), this)
		registerApplicationMessages.registerForClass(cls<PlaybackMessage.PlaybackStopped>(), this)
		registerApplicationMessages.registerForClass(cls<FilePropertiesUpdatedMessage>(), this)
	}

	@Volatile
	private var activeNowPlayingUrlKey: UrlKeyHolder<ServiceFile>? = null

	override fun invoke(message: ApplicationMessage) {
		when (message) {
			is PlaybackMessage.TrackChanged -> updatePlayingFilePropertiesIfNecessary()
			is PlaybackMessage.PlaybackStarted -> playbackNotificationBroadcaster.notifyPlaying()
			is PlaybackMessage.PlaybackPaused -> playbackNotificationBroadcaster.notifyPaused()
			is PlaybackMessage.PlaybackInterrupted -> playbackNotificationBroadcaster.notifyInterrupted()
			is PlaybackMessage.PlaybackStopped -> playbackNotificationBroadcaster.notifyStopped()
			is FilePropertiesUpdatedMessage -> updatePlayingFilePropertiesIfNecessary()
		}
	}

	override fun close() {
		registerApplicationMessages.unregisterReceiver(this)
	}

	private fun updatePlayingFilePropertiesIfNecessary() {
		nowPlayingProvider
			.promiseNowPlaying()
			.then { nowPlaying ->
				nowPlaying?.playingFile?.serviceFile?.also {
					scopedUrlKeys
						.promiseUrlKey(it)
						.then { urlKeyHolder ->
							if (activeNowPlayingUrlKey != urlKeyHolder) {
								activeNowPlayingUrlKey = urlKeyHolder
								playbackNotificationBroadcaster.notifyPlayingFileUpdated()
							}
						}
				}
			}
	}
}
